### ISSUE TITLE
Optimizing Autoencoder with optuna

### ISSUE THREAD

**Issue by arilwan** [optuna/optuna#2555]
I am trying to optimize an `autoencoder` network with `optuna`, but I am having difficulty deciding an appropriate objective function return value. As we know, the goal in `autoencoder` is to reconstruct the input, so initially, I thought my objective function shall focus on minimizing the reconstruction error (`mse` in this case).  To do this, I write the following objective function:

```
def objective(trial):
    """Object function to optimize parameters"""
    #clear session
    keras.backend.clear_session()

    encoding_dim = trial.suggest_int('mid_units', 32, 180)
    input_shape  = x_train_sub.shape[1:]  #(1, 100, 4) the shape of instances
    autoencoder  = Sequential()
    activ='relu'

    autoencoder.add(Flatten(input_shape=input_shape))
    autoencoder.add(Dense(600, activation=activ))#, kernel_regularizer=l2(0.01)))
    autoencoder.add(Dense(400, activation=activ))#, kernel_regularizer=l2(0.01)))
    autoencoder.add(Dense(200, activation=activ))
    autoencoder.add(Dense(encoding_dim, activation=activ))   

    #decoder
    autoencoder.add(Dense(200, activation=activ))
    autoencoder.add(Dense(400, activation=activ))
    autoencoder.add(Dense(600, activation=activ))
    autoencoder.add(Dense(np.prod(input_shape), activation=activ)) 
    autoencoder.add(Reshape(input_shape))

    optimizer = Adam(lr=trial.suggest_loguniform("learning_rate", 1e-5, 1e-1)) 
    autoencoder.compile(optimizer=optimizer, loss='mae', metrics=['mean_squared_error'])#metrics=['mean_squared_error'])
    hist = autoencoder.fit(x_train_sub, x_train_sub, epochs=20, verbose = 0,callbacks=[KerasPruningCallback(trial, "loss")],
        batch_size = trial.suggest_categorical('batch_size', [64, 128, 256]), shuffle=True)

    min_mse = np.min(hist.history['mean_squared_error'])

    return min_mse

```
However, this doesn't optimize network performance, just similar to using `autoencoder` with `optuna`. I suspect my return value isn't correctly defined. Can someone help with this?

**Comment by nzw0301** [comment]
hi, sorry for my unclear answer on StackOverflow.

I think the objective should return a mean squared error on a validatiaon dataset rather than the training dataset `x_train_sub`. What we really want is the auto-encoder model can reconstruct input data in an **unseen dataset**, called test dataset. But we cannot observe the test dataset in practice, we need to prepare a validation dataset that is a disjoint set of the training dataset to approximate the test dataset. 

I suppose that there are a number of reasons why optuna cannot improve the performance of the auto-encoder. 

1. the number of trials is not enough
2. searched space of hyper-parameter is limited; we can enlarge the searched space, for example, the number of layers, the number of hidden units, and optimizer
3. the task is too easy to solve. the hand-tuned hyper-parameters already give a performance.

**Comment by arilwan** [comment]
@nzw0301  Ah, OK. So you recommend that I set aside a `validation set`. How about the `mean_squared_error`  because the `fit()` a list of ´mse` so I was trying to retrieve the minimum error from the log and return such. Is that a way to go?

**Comment by nzw0301** [comment]
I think that is fine; it sounds equivalent to perform early-stopping.

**Comment by arilwan** [comment]
many thanks.
